File: apps/mobile/ios/App/App/CompanionAuthPlugin.swift
import Foundation
import Security
import Capacitor

private final class NoRedirectSessionDelegate: NSObject, URLSessionTaskDelegate {
    func urlSession(_ session: URLSession, task: URLSessionTask, willPerformHTTPRedirection response: HTTPURLResponse,
                    newRequest request: URLRequest, completionHandler: @escaping (URLRequest?) -> Void) {
        completionHandler(nil)
    }
}

private final class RefreshOperation {
    let generation: UInt64
    let origin: String
    var calls: [CAPPluginCall]

    init(generation: UInt64, origin: String, call: CAPPluginCall) {
        self.generation = generation
        self.origin = origin
        self.calls = [call]
    }
}

@objc(CompanionAuthPlugin)
public class CompanionAuthPlugin: CAPPlugin, CAPBridgedPlugin {
    public let identifier = "CompanionAuthPlugin"
    public let jsName = "CompanionAuth"
    public let pluginMethods: [CAPPluginMethod] = [
        CAPPluginMethod(name: "persistAuth", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "getRefreshToken", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "getAccessToken", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "refreshAccessToken", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "clearAccessToken", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "clearAuth", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "returnToLauncher", returnType: CAPPluginReturnPromise),
    ]

    private let service = "space.companion.mobile.auth"
    private let refreshAccount = "refresh-token-v1"
    private let rotationAccount = "rotation-id-v1"
    private var accessToken: String?
    private var accessTokenExpiresAt: String?
    private var authorizedOrigin: String?
    private let lock = NSLock()
    private var authGeneration: UInt64 = 0
    private var refreshOperation: RefreshOperation?

    @objc func persistAuth(_ call: CAPPluginCall) {
        guard isLocalLauncher else {
            call.reject("Mobile auth persistence is restricted to the local launcher")
            return
        }
        guard let refreshToken = call.getString("refreshToken"), refreshToken.count >= 32,
              let nextAccessToken = call.getString("accessToken"),
              let expiresAt = call.getString("accessTokenExpiresAt"),
              let serverOrigin = call.getString("serverOrigin"), MobileAuthPolicy.isValidServerOrigin(serverOrigin),
              Self.parseISODate(expiresAt) != nil else {
            call.reject("Invalid mobile auth payload")
            return
        }
        do {
            let staleCalls: [CAPPluginCall]
            lock.lock()
            do {
                try writeKeychain(account: refreshAccount, value: refreshToken)
                if let rotationId = call.getString("rotationId"), !rotationId.isEmpty {
                    try writeKeychain(account: rotationAccount, value: rotationId)
                } else {
                    deleteKeychain(account: rotationAccount)
                }
                staleCalls = invalidateRefreshLocked()
                accessToken = nextAccessToken
                accessTokenExpiresAt = expiresAt
                authorizedOrigin = serverOrigin
                lock.unlock()
            } catch {
                lock.unlock()
                throw error
            }
            reject(calls: staleCalls, message: "Mobile auth changed while access was refreshing", code: "AUTH_REVOKED")
            call.resolve()
        } catch {
            call.reject("Unable to store mobile auth", nil, error)
        }
    }

    @objc func getRefreshToken(_ call: CAPPluginCall) {
        guard isLocalLauncher else {
            call.reject("Refresh credentials are restricted to the local launcher")
            return
        }
        do {
            let refresh = try readKeychain(account: refreshAccount)
            let rotation = try readKeychain(account: rotationAccount)
            call.resolve(["value": refresh.map { $0 as Any } ?? NSNull(),
                          "rotationId": rotation.map { $0 as Any } ?? NSNull()])
        } catch {
            clearStoredAuth()
            call.reject("Unable to read mobile auth", nil, error)
        }
    }

    @objc func getAccessToken(_ call: CAPPluginCall) {
        lock.lock()
        guard isAuthorizedRemotePageLocked(origin: authorizedOrigin) else {
            lock.unlock()
            call.reject("Access token is restricted to the authorized server")
            return
        }
        let expiry = accessTokenExpiresAt.flatMap { Self.parseISODate($0) }
        guard let token = accessToken, let expiresAt = accessTokenExpiresAt, let expiry, expiry > Date() else {
            let staleCalls = clearAccessTokenStateLocked(clearOrigin: false)
            lock.unlock()
            reject(calls: staleCalls, message: "Mobile access refresh was cancelled", code: "AUTH_REVOKED")
            call.resolve(["value": NSNull(), "expiresAt": NSNull()])
            return
        }
        lock.unlock()
        call.resolve(["value": token, "expiresAt": expiresAt])
    }

    @objc func refreshAccessToken(_ call: CAPPluginCall) {
        lock.lock()
        guard let approvedOrigin = authorizedOrigin, isAuthorizedRemotePageLocked(origin: approvedOrigin) else {
            lock.unlock()
            call.reject("Access refresh is restricted to the authorized server", "AUTH_REVOKED")
            return
        }
        if let active = refreshOperation, active.generation == authGeneration, active.origin == approvedOrigin {
            active.calls.append(call)
            lock.unlock()
            return
        }
        let operation = RefreshOperation(generation: authGeneration, origin: approvedOrigin, call: call)
        refreshOperation = operation
        let refreshToken: String
        do {
            guard let stored = try readKeychain(account: refreshAccount), stored.count >= 32 else {
                deleteKeychain(account: refreshAccount)
                deleteKeychain(account: rotationAccount)
                let calls = clearAccessTokenStateLocked(clearOrigin: true)
                lock.unlock()
                returnToLocalLauncher()
                reject(calls: calls, message: "Mobile refresh credential is missing", code: "AUTH_REVOKED")
                return
            }
            refreshToken = stored
        } catch {
            let calls = detachCallsLocked(operation)
            lock.unlock()
            reject(calls: calls, message: "Mobile access refresh is temporarily unavailable", code: "AUTH_RETRYABLE", error: error)
            return
        }
        lock.unlock()
        guard var endpoint = URLComponents(string: approvedOrigin) else {
            finish(operation: operation, message: "Unable to construct mobile refresh request", code: "AUTH_RETRYABLE")
            return
        }
        endpoint.path = "/api/v1/mobile/auth/refresh"
        endpoint.query = nil
        endpoint.fragment = nil
        guard let url = endpoint.url,
              let body = try? JSONSerialization.data(withJSONObject: ["refresh_token": refreshToken]) else {
            finish(operation: operation, message: "Unable to construct mobile refresh request", code: "AUTH_RETRYABLE")
            return
        }
        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.httpBody = body
        request.timeoutInterval = 10
        request.cachePolicy = .reloadIgnoringLocalCacheData
        request.setValue("application/json", forHTTPHeaderField: "Accept")
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")

        let configuration = URLSessionConfiguration.ephemeral
        configuration.timeoutIntervalForRequest = 10
        configuration.timeoutIntervalForResource = 10
        configuration.httpCookieStorage = nil
        configuration.urlCache = nil
        let session = URLSession(configuration: configuration, delegate: NoRedirectSessionDelegate(), delegateQueue: nil)
        session.dataTask(with: request) { [weak self] data, response, error in
            defer { session.finishTasksAndInvalidate() }
            guard let self else { return }
            if let error {
                self.finish(operation: operation, message: "Mobile access refresh is temporarily unavailable", code: "AUTH_RETRYABLE", error: error)
                return
            }
            guard let http = response as? HTTPURLResponse else {
                self.finish(operation: operation, message: "Mobile access refresh returned no HTTP response", code: "AUTH_RETRYABLE")
                return
            }
            if http.statusCode == 401 {
                let calls = self.revokeAndReturnToLauncherIfCurrent(operation)
                self.reject(calls: calls, message: "Mobile refresh credential was revoked", code: "AUTH_REVOKED")
                return
            }
            guard http.statusCode == 200 else {
                self.finish(operation: operation, message: "Mobile access refresh failed with HTTP \(http.statusCode)", code: "AUTH_RETRYABLE")
                return
            }
            guard let data, data.count <= 64 * 1024,
                  let json = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
                  let nextRefresh = json["refresh_token"] as? String, nextRefresh.count >= 32,
                  let nextAccess = json["access_token"] as? String, nextAccess.count >= 20,
                  let expiresAt = json["access_token_expires_at"] as? String,
                  let expiry = Self.parseISODate(expiresAt), expiry > Date() else {
                self.finish(operation: operation, message: "Mobile access refresh returned an invalid response", code: "AUTH_RETRYABLE")
                return
            }
            self.lock.lock()
            guard self.isCurrentRefreshLocked(operation) else {
                let calls = self.detachCallsLocked(operation)
                self.lock.unlock()
                self.reject(calls: calls, message: "Mobile auth changed while access was refreshing", code: "AUTH_REVOKED")
                return
            }
            do {
                try self.writeKeychain(account: self.refreshAccount, value: nextRefresh)
                self.accessToken = nextAccess
                self.accessTokenExpiresAt = expiresAt
                let calls = self.detachCallsLocked(operation)
                for pending in calls { pending.resolve(["value": nextAccess, "expiresAt": expiresAt]) }
                self.lock.unlock()
            } catch {
                let calls = self.detachCallsLocked(operation)
                self.lock.unlock()
                self.reject(calls: calls, message: "Unable to persist rotated mobile credential", code: "AUTH_RETRYABLE", error: error)
            }
        }.resume()
    }

    @objc func clearAccessToken(_ call: CAPPluginCall) {
        guard isAuthorizedRemotePage else {
            call.reject("Access-token clearing is restricted to the authorized server")
            return
        }
        let staleCalls = clearAccessTokenState(clearOrigin: true)
        reject(calls: staleCalls, message: "Mobile access refresh was cancelled", code: "AUTH_REVOKED")
        call.resolve()
    }

    @objc func returnToLauncher(_ call: CAPPluginCall) {
        guard isAuthorizedRemotePage, let localURL = bridge?.config.localURL else {
            call.reject("Launcher return is restricted to the authorized server")
            return
        }
        let staleCalls = clearAccessTokenState(clearOrigin: true)
        reject(calls: staleCalls, message: "Mobile access refresh was cancelled", code: "AUTH_REVOKED")
        DispatchQueue.main.async { [weak self] in
            self?.bridge?.webView?.load(URLRequest(url: localURL))
        }
        call.resolve()
    }

    @objc func clearAuth(_ call: CAPPluginCall) {
        guard isLocalLauncher else {
            call.reject("Full mobile unpairing is restricted to the local launcher")
            return
        }
        clearStoredAuth()
        call.resolve()
    }

    private func keychainQuery(account: String) -> [String: Any] {
        [kSecClass as String: kSecClassGenericPassword,
         kSecAttrService as String: service,
         kSecAttrAccount as String: account]
    }

    private var isLocalLauncher: Bool {
        bridge?.webView?.url?.scheme == "capacitor" && bridge?.webView?.url?.host == "app.companion.local"
    }

    private var isAuthorizedRemotePage: Bool {
        lock.lock()
        defer { lock.unlock() }
        return isAuthorizedRemotePageLocked(origin: authorizedOrigin)
    }

    private func isAuthorizedRemotePageLocked(origin: String?) -> Bool {
        MobileAuthPolicy.isSameOrigin(current: bridge?.webView?.url, approvedText: origin)
    }

    private func writeKeychain(account: String, value: String) throws {
        let query = keychainQuery(account: account)
        let data = Data(value.utf8)
        let updateStatus = SecItemUpdate(query as CFDictionary, [kSecValueData as String: data] as CFDictionary)
        if updateStatus == errSecSuccess { return }
        guard updateStatus == errSecItemNotFound else {
            throw NSError(domain: NSOSStatusErrorDomain, code: Int(updateStatus))
        }
        var insert = query
        insert[kSecValueData as String] = data
        insert[kSecAttrAccessible as String] = kSecAttrAccessibleAfterFirstUnlockThisDeviceOnly
        let insertStatus = SecItemAdd(insert as CFDictionary, nil)
        guard insertStatus == errSecSuccess else { throw NSError(domain: NSOSStatusErrorDomain, code: Int(insertStatus)) }
    }

    private func readKeychain(account: String) throws -> String? {
        var query = keychainQuery(account: account)
        query[kSecReturnData as String] = true
        query[kSecMatchLimit as String] = kSecMatchLimitOne
        var result: CFTypeRef?
        let status = SecItemCopyMatching(query as CFDictionary, &result)
        if status == errSecItemNotFound { return nil }
        guard status == errSecSuccess, let data = result as? Data, let value = String(data: data, encoding: .utf8) else {
            throw NSError(domain: NSOSStatusErrorDomain, code: Int(status))
        }
        return value
    }

    private func deleteKeychain(account: String) {
        SecItemDelete(keychainQuery(account: account) as CFDictionary)
    }

    private func clearStoredAuth() {
        lock.lock()
        deleteKeychain(account: refreshAccount)
        deleteKeychain(account: rotationAccount)
        let calls = clearAccessTokenStateLocked(clearOrigin: true)
        lock.unlock()
        reject(calls: calls, message: "Mobile auth was cleared", code: "AUTH_REVOKED")
    }

    private func revokeAndReturnToLauncherIfCurrent(_ operation: RefreshOperation) -> [CAPPluginCall] {
        lock.lock()
        guard isCurrentRefreshLocked(operation) else {
            let calls = detachCallsLocked(operation)
            lock.unlock()
            return calls
        }
        deleteKeychain(account: refreshAccount)
        deleteKeychain(account: rotationAccount)
        let calls = clearAccessTokenStateLocked(clearOrigin: true)
        lock.unlock()
        returnToLocalLauncher()
        return calls
    }

    private func returnToLocalLauncher() {
        guard let localURL = bridge?.config.localURL else { return }
        DispatchQueue.main.async { [weak self] in
            self?.bridge?.webView?.load(URLRequest(url: localURL))
        }
    }

    private static func parseISODate(_ value: String) -> Date? {
        let standard = ISO8601DateFormatter()
        if let parsed = standard.date(from: value) { return parsed }
        let fractional = ISO8601DateFormatter()
        fractional.formatOptions = [.withInternetDateTime, .withFractionalSeconds]
        return fractional.date(from: value)
    }

    private func clearAccessTokenState(clearOrigin: Bool) -> [CAPPluginCall] {
        lock.lock()
        let calls = clearAccessTokenStateLocked(clearOrigin: clearOrigin)
        lock.unlock()
        return calls
    }

    private func clearAccessTokenStateLocked(clearOrigin: Bool) -> [CAPPluginCall] {
        let calls = invalidateRefreshLocked()
        accessToken = nil
        accessTokenExpiresAt = nil
        if clearOrigin { authorizedOrigin = nil }
        return calls
    }

    private func invalidateRefreshLocked() -> [CAPPluginCall] {
        authGeneration &+= 1
        guard let active = refreshOperation else { return [] }
        refreshOperation = nil
        let calls = active.calls
        active.calls.removeAll()
        return calls
    }

    private func isCurrentRefreshLocked(_ operation: RefreshOperation) -> Bool {
        refreshOperation === operation && MobileAuthPolicy.isCurrentRefresh(
            operationGeneration: operation.generation,
            currentGeneration: authGeneration,
            operationOrigin: operation.origin,
            authorizedOrigin: authorizedOrigin,
            currentURL: bridge?.webView?.url
        )
    }

    private func detachCallsLocked(_ operation: RefreshOperation) -> [CAPPluginCall] {
        if refreshOperation === operation { refreshOperation = nil }
        let calls = operation.calls
        operation.calls.removeAll()
        return calls
    }

    private func finish(operation: RefreshOperation, message: String, code: String, error: Error? = nil) {
        lock.lock()
        let calls = detachCallsLocked(operation)
        lock.unlock()
        reject(calls: calls, message: message, code: code, error: error)
    }

    private func reject(calls: [CAPPluginCall], message: String, code: String, error: Error? = nil) {
        for call in calls { call.reject(message, code, error) }
    }
}

File: apps/mobile/android/app/src/main/java/space/companion/mobile/CompanionAuthPlugin.java
package space.companion.mobile;

import android.content.Context;
import android.content.SharedPreferences;
import android.security.keystore.KeyGenParameterSpec;
import android.security.keystore.KeyProperties;
import android.util.Base64;

import com.getcapacitor.JSObject;
import com.getcapacitor.Plugin;
import com.getcapacitor.PluginCall;
import com.getcapacitor.PluginMethod;
import com.getcapacitor.annotation.CapacitorPlugin;

import org.json.JSONObject;

import java.io.ByteArrayOutputStream;
import java.io.InputStream;
import java.io.OutputStream;
import java.net.HttpURLConnection;
import java.net.URL;
import java.nio.charset.StandardCharsets;
import java.security.KeyStore;
import java.util.ArrayList;
import java.util.List;
import java.util.Objects;

import javax.crypto.Cipher;
import javax.crypto.KeyGenerator;
import javax.crypto.SecretKey;
import javax.crypto.spec.GCMParameterSpec;

@CapacitorPlugin(name = "CompanionAuth")
public class CompanionAuthPlugin extends Plugin {
    private static final int HTTP_TIMEOUT_MS = 10_000;
    private static final int MAX_AUTH_RESPONSE_BYTES = 64 * 1024;
    private static final String KEY_ALIAS = "companion_space_mobile_refresh_v1";
    private static final String PREFS = "companion_space_mobile_secure_auth";
    private static final String REFRESH = "refresh";
    private static final String ROTATION = "rotation";
    private volatile String accessToken;
    private volatile String accessTokenExpiresAt;
    private volatile long accessTokenExpiresAtEpochMs;
    private volatile String authorizedOrigin;
    private final Object authLock = new Object();
    private final AuthRefreshGate<RefreshOperation> refreshGate = new AuthRefreshGate<>();

    private static final class RefreshOperation {
        final long generation;
        final String origin;
        final List<PluginCall> calls = new ArrayList<>();

        RefreshOperation(long generation, String origin, PluginCall call) {
            this.generation = generation;
            this.origin = origin;
            calls.add(call);
        }
    }


    @PluginMethod
    public void persistAuth(PluginCall call) {
        if (!isLocalLauncher()) {
            call.reject("Mobile auth persistence is restricted to the local launcher");
            return;
        }
        String refreshToken = call.getString("refreshToken");
        String nextAccessToken = call.getString("accessToken");
        String expiresAt = call.getString("accessTokenExpiresAt");
        Long expiresAtEpochMs = call.getLong("accessTokenExpiresAtEpochMs");
        String rotationId = call.getString("rotationId");
        String serverOrigin = call.getString("serverOrigin");
        if (refreshToken == null || refreshToken.length() < 32 || nextAccessToken == null || expiresAt == null || expiresAtEpochMs == null || !isValidHttpsOrigin(serverOrigin)) {
            call.reject("Invalid mobile auth payload");
            return;
        }
        try {
            String encryptedRefresh = encrypt(refreshToken);
            String encryptedRotation = rotationId == null || rotationId.isBlank() ? null : encrypt(rotationId);
            RefreshOperation stale;
            synchronized (authLock) {
                SharedPreferences.Editor editor = prefs().edit().putString(REFRESH, encryptedRefresh);
                if (encryptedRotation == null) editor.remove(ROTATION);
                else editor.putString(ROTATION, encryptedRotation);
                if (!editor.commit()) throw new IllegalStateException("Secure auth commit failed");
                stale = invalidateRefreshLocked();
                accessToken = nextAccessToken;
                accessTokenExpiresAt = expiresAt;
                accessTokenExpiresAtEpochMs = expiresAtEpochMs;
                authorizedOrigin = serverOrigin;
            }
            rejectCalls(stale, "Mobile auth changed while access was refreshing", "AUTH_REVOKED");
            call.resolve();
        } catch (Exception error) {
            call.reject("Unable to store mobile auth", error);
        }
    }

    @PluginMethod
    public void getRefreshToken(PluginCall call) {
        if (!isLocalLauncher()) {
            call.reject("Refresh credentials are restricted to the local launcher");
            return;
        }
        JSObject result = new JSObject();
        try {
            result.put("value", decryptNullable(prefs().getString(REFRESH, null)));
            result.put("rotationId", decryptNullable(prefs().getString(ROTATION, null)));
            call.resolve(result);
        } catch (Exception error) {
            clearStoredAuth();
            call.reject("Unable to read mobile auth", error);
        }
    }

    @PluginMethod
    public void getAccessToken(PluginCall call) {
        JSObject result = new JSObject();
        RefreshOperation stale = null;
        boolean unauthorized;
        synchronized (authLock) {
            unauthorized = !isAuthorizedRemotePageLocked(authorizedOrigin);
            if (!unauthorized && accessToken != null && accessTokenExpiresAt != null && accessTokenExpiresAtEpochMs > System.currentTimeMillis()) {
                result.put("value", accessToken);
                result.put("expiresAt", accessTokenExpiresAt);
            } else if (!unauthorized) {
                stale = clearAccessTokenStateLocked(false);
                result.put("value", JSObject.NULL);
                result.put("expiresAt", JSObject.NULL);
            }
        }
        if (unauthorized) {
            call.reject("Access token is restricted to the authorized server");
            return;
        }
        rejectCalls(stale, "Mobile access refresh was cancelled", "AUTH_REVOKED");
        call.resolve(result);
    }

    @PluginMethod
    public void refreshAccessToken(PluginCall call) {
        if (!isAuthorizedRemotePage()) {
            call.reject("Access refresh is restricted to the authorized server", "AUTH_REVOKED");
            return;
        }
        final RefreshOperation operation;
        synchronized (authLock) {
            String approvedOrigin = authorizedOrigin;
            if (!isAuthorizedRemotePageLocked(approvedOrigin)) {
                call.reject("Access refresh is restricted to the authorized server", "AUTH_REVOKED");
                return;
            }
            RefreshOperation active = refreshGate.active();
            if (active != null
                    && active.generation == refreshGate.generation()
                    && Objects.equals(active.origin, approvedOrigin)) {
                active.calls.add(call);
                return;
            }
            operation = new RefreshOperation(refreshGate.generation(), approvedOrigin, call);
            refreshGate.setActive(operation);
        }
        execute(() -> refreshAccessTokenInBackground(operation));
    }

    @PluginMethod
    public void clearAccessToken(PluginCall call) {
        if (!isAuthorizedRemotePage()) {
            call.reject("Access-token clearing is restricted to the authorized server");
            return;
        }
        RefreshOperation stale = clearAccessTokenState(true);
        rejectCalls(stale, "Mobile access refresh was cancelled", "AUTH_REVOKED");
        call.resolve();
    }

    @PluginMethod
    public void returnToLauncher(PluginCall call) {
        if (!isAuthorizedRemotePage()) {
            call.reject("Launcher return is restricted to the authorized server");
            return;
        }
        RefreshOperation stale = clearAccessTokenState(true);
        rejectCalls(stale, "Mobile access refresh was cancelled", "AUTH_REVOKED");
        getBridge().getWebView().post(() -> getBridge().getWebView().loadUrl(getBridge().getLocalUrl()));
        call.resolve();
    }

    @PluginMethod
    public void clearAuth(PluginCall call) {
        if (!isLocalLauncher()) {
            call.reject("Full mobile unpairing is restricted to the local launcher");
            return;
        }
        clearStoredAuth();
        call.resolve();
    }

    private SharedPreferences prefs() {
        return getContext().getSharedPreferences(PREFS, Context.MODE_PRIVATE);
    }

    private boolean isLocalLauncher() {
        String currentUrl = getBridge().getWebView().getUrl();
        return currentUrl != null && currentUrl.startsWith("https://app.companion.local/");
    }

    private boolean isValidHttpsOrigin(String value) {
        return MobileOriginPolicy.isValidServerOrigin(value);
    }

    private boolean isAuthorizedRemotePage() {
        synchronized (authLock) {
            return isAuthorizedRemotePageLocked(authorizedOrigin);
        }
    }

    private boolean isAuthorizedRemotePageLocked(String approvedOrigin) {
        String currentUrl = getBridge().getWebView().getUrl();
        return currentUrl != null && approvedOrigin != null && MobileOriginPolicy.isSameOrigin(approvedOrigin, currentUrl);
    }

    private void refreshAccessTokenInBackground(RefreshOperation operation) {
        HttpURLConnection connection = null;
        try {
            String refreshToken = null;
            List<PluginCall> staleCalls = null;
            synchronized (authLock) {
                if (!isCurrentRefreshLocked(operation)) {
                    staleCalls = detachCallsLocked(operation);
                } else {
                    refreshToken = decryptNullable(prefs().getString(REFRESH, null));
                }
            }
            if (staleCalls != null) {
                rejectCalls(staleCalls, "Mobile auth changed while access was refreshing", "AUTH_REVOKED", null);
                return;
            }
            if (refreshToken == null || refreshToken.length() < 32) {
                List<PluginCall> calls = revokeAndReturnToLauncherIfCurrent(operation);
                rejectCalls(calls, "Mobile refresh credential is missing", "AUTH_REVOKED", null);
                return;
            }
            URL endpoint = new URL(operation.origin + "/api/v1/mobile/auth/refresh");
            connection = (HttpURLConnection) endpoint.openConnection();
            connection.setInstanceFollowRedirects(false);
            connection.setConnectTimeout(HTTP_TIMEOUT_MS);
            connection.setReadTimeout(HTTP_TIMEOUT_MS);
            connection.setRequestMethod("POST");
            connection.setDoOutput(true);
            connection.setRequestProperty("Accept", "application/json");
            connection.setRequestProperty("Content-Type", "application/json");
            byte[] requestBody = new JSONObject().put("refresh_token", refreshToken).toString().getBytes(StandardCharsets.UTF_8);
            connection.setFixedLengthStreamingMode(requestBody.length);
            try (OutputStream output = connection.getOutputStream()) {
                output.write(requestBody);
            }
            refreshToken = null;

            int status = connection.getResponseCode();
            if (status == HttpURLConnection.HTTP_UNAUTHORIZED) {
                List<PluginCall> calls = revokeAndReturnToLauncherIfCurrent(operation);
                rejectCalls(calls, "Mobile refresh credential was revoked", "AUTH_REVOKED", null);
                return;
            }
            if (status != HttpURLConnection.HTTP_OK) {
                finishWithError(operation, "Mobile access refresh failed with HTTP " + status, "AUTH_RETRYABLE", null);
                return;
            }
            String responseText = readLimited(connection.getInputStream());
            JSONObject response = new JSONObject(responseText);
            String nextRefresh = response.optString("refresh_token", "");
            String nextAccess = response.optString("access_token", "");
            String expiresAt = response.optString("access_token_expires_at", "");
            long expiresAtEpochMs = MobileOriginPolicy.parseIsoEpochMs(expiresAt);
            if (nextRefresh.length() < 32 || nextAccess.length() < 20 || expiresAtEpochMs <= System.currentTimeMillis()) {
                finishWithError(operation, "Mobile access refresh returned an invalid response", "AUTH_RETRYABLE", null);
                return;
            }
            String encryptedRefresh = encrypt(nextRefresh);
            nextRefresh = null;
            JSObject result = new JSObject();
            result.put("value", nextAccess);
            result.put("expiresAt", expiresAt);
            synchronized (authLock) {
                if (!isCurrentRefreshLocked(operation)) {
                    rejectCalls(detachCallsLocked(operation), "Mobile auth changed while access was refreshing", "AUTH_REVOKED", null);
                    return;
                }
                if (!prefs().edit().putString(REFRESH, encryptedRefresh).commit()) {
                    rejectCalls(detachCallsLocked(operation), "Unable to persist rotated mobile credential", "AUTH_RETRYABLE", null);
                    return;
                }
                accessToken = nextAccess;
                accessTokenExpiresAt = expiresAt;
                accessTokenExpiresAtEpochMs = expiresAtEpochMs;
                resolveCalls(detachCallsLocked(operation), result);
            }
        } catch (Exception error) {
            finishWithError(operation, "Mobile access refresh is temporarily unavailable", "AUTH_RETRYABLE", error);
        } finally {
            if (connection != null) connection.disconnect();
        }
    }

    private String readLimited(InputStream input) throws Exception {
        try (InputStream stream = input; ByteArrayOutputStream output = new ByteArrayOutputStream()) {
            byte[] buffer = new byte[4096];
            int total = 0;
            int read;
            while ((read = stream.read(buffer)) != -1) {
                total += read;
                if (total > MAX_AUTH_RESPONSE_BYTES) throw new IllegalStateException("Auth response too large");
                output.write(buffer, 0, read);
            }
            return output.toString(StandardCharsets.UTF_8.name());
        }
    }

    private List<PluginCall> revokeAndReturnToLauncherIfCurrent(RefreshOperation operation) {
        List<PluginCall> calls;
        synchronized (authLock) {
            if (!isCurrentRefreshLocked(operation)) return detachCallsLocked(operation);
            prefs().edit().clear().commit();
            calls = drainCalls(clearAccessTokenStateLocked(true));
        }
        getBridge().getWebView().post(() -> getBridge().getWebView().loadUrl(getBridge().getLocalUrl()));
        return calls;
    }

    private void clearStoredAuth() {
        RefreshOperation stale;
        synchronized (authLock) {
            prefs().edit().clear().commit();
            stale = clearAccessTokenStateLocked(true);
        }
        rejectCalls(stale, "Mobile auth was cleared", "AUTH_REVOKED");
    }

    private RefreshOperation clearAccessTokenState(boolean clearOrigin) {
        synchronized (authLock) {
            return clearAccessTokenStateLocked(clearOrigin);
        }
    }

    private RefreshOperation clearAccessTokenStateLocked(boolean clearOrigin) {
        RefreshOperation stale = invalidateRefreshLocked();
        accessToken = null;
        accessTokenExpiresAt = null;
        accessTokenExpiresAtEpochMs = 0;
        if (clearOrigin) authorizedOrigin = null;
        return stale;
    }

    private RefreshOperation invalidateRefreshLocked() {
        return refreshGate.invalidate();
    }

    private boolean isCurrentRefreshLocked(RefreshOperation operation) {
        return refreshGate.isCurrent(operation.generation, operation)
                && Objects.equals(authorizedOrigin, operation.origin)
                && isAuthorizedRemotePageLocked(operation.origin);
    }

    private List<PluginCall> detachCallsLocked(RefreshOperation operation) {
        refreshGate.detach(operation);
        List<PluginCall> calls = new ArrayList<>(operation.calls);
        operation.calls.clear();
        return calls;
    }

    private void finishWithError(RefreshOperation operation, String message, String code, Exception error) {
        List<PluginCall> calls;
        synchronized (authLock) {
            calls = detachCallsLocked(operation);
        }
        rejectCalls(calls, message, code, error);
    }

    private void resolveCalls(List<PluginCall> calls, JSObject result) {
        for (PluginCall call : calls) call.resolve(result);
    }

    private void rejectCalls(RefreshOperation operation, String message, String code) {
        rejectCalls(drainCalls(operation), message, code, null);
    }

    private List<PluginCall> drainCalls(RefreshOperation operation) {
        if (operation == null) return new ArrayList<>();
        List<PluginCall> calls = new ArrayList<>(operation.calls);
        operation.calls.clear();
        return calls;
    }

    private void rejectCalls(List<PluginCall> calls, String message, String code, Exception error) {
        for (PluginCall call : calls) {
            if (error == null) call.reject(message, code);
            else call.reject(message, code, error);
        }
    }

    private SecretKey secretKey() throws Exception {
        KeyStore store = KeyStore.getInstance("AndroidKeyStore");
        store.load(null);
        if (store.containsAlias(KEY_ALIAS)) return ((KeyStore.SecretKeyEntry) store.getEntry(KEY_ALIAS, null)).getSecretKey();
        KeyGenerator generator = KeyGenerator.getInstance(KeyProperties.KEY_ALGORITHM_AES, "AndroidKeyStore");
        generator.init(new KeyGenParameterSpec.Builder(KEY_ALIAS, KeyProperties.PURPOSE_ENCRYPT | KeyProperties.PURPOSE_DECRYPT)
                .setBlockModes(KeyProperties.BLOCK_MODE_GCM)
                .setEncryptionPaddings(KeyProperties.ENCRYPTION_PADDING_NONE)
                .setRandomizedEncryptionRequired(true)
                .build());
        return generator.generateKey();
    }

    private String encrypt(String value) throws Exception {
        Cipher cipher = Cipher.getInstance("AES/GCM/NoPadding");
        cipher.init(Cipher.ENCRYPT_MODE, secretKey());
        String iv = Base64.encodeToString(cipher.getIV(), Base64.NO_WRAP);
        String body = Base64.encodeToString(cipher.doFinal(value.getBytes(StandardCharsets.UTF_8)), Base64.NO_WRAP);
        return iv + "." + body;
    }

    private String decryptNullable(String encoded) throws Exception {
        if (encoded == null) return null;
        String[] parts = encoded.split("\\.", 2);
        if (parts.length != 2) throw new IllegalStateException("Invalid encrypted auth payload");
        Cipher cipher = Cipher.getInstance("AES/GCM/NoPadding");
        cipher.init(Cipher.DECRYPT_MODE, secretKey(), new GCMParameterSpec(128, Base64.decode(parts[0], Base64.NO_WRAP)));
        return new String(cipher.doFinal(Base64.decode(parts[1], Base64.NO_WRAP)), StandardCharsets.UTF_8);
    }
}

final class AuthRefreshGate<T> {
    private long generation;
    private T active;

    long generation() {
        return generation;
    }

    T active() {
        return active;
    }

    void setActive(T value) {
        active = value;
    }

    T invalidate() {
        generation++;
        T stale = active;
        active = null;
        return stale;
    }

    boolean isCurrent(long candidateGeneration, T candidate) {
        return generation == candidateGeneration && active == candidate;
    }

    void detach(T candidate) {
        if (active == candidate) active = null;
    }
}
